core/trino-main/src/main/java/io/trino/sql/rewrite/ShowQueriesRewrite.java

```diff
@@ -397,6 +397,10 @@ protected Node visitShowSchemas(ShowSchemas node, Void context)
             }
 
             String catalog = node.getCatalog().map(Identifier::getValue).orElseGet(() -> session.getCatalog().orElseThrow());
+            if (!metadata.catalogExists(session, catalog)) {
```

Comment:
Does this change the security behavior in the case where `io.trino.spi.security.SystemAccessControl#canAccessCatalog` returns false?

I see that method is called from every method in `AccessControlManager`, so the below `checkCanShowSchemas()` would fail with `Cannot access catalog` if the user doesn't have access. After this change, would `SHOW SCHEMAS` start failing with `CATALOG_NOT_FOUND` instead of `ACCESS_DENIED`?

Reply:
Right now the code is inconsistent in different `SHOW CREATE` cases (i.e. `SHOW CREATE SCHEMA` checks for schema existence. I agree that the `SHOW CREATE` statements shouldn't reveal whether relation exists (even if the user has access to it) so I think that in both cases - relation does not exist or user has no access to it, we should throw not found exception. WDYT?

Reply:
@dain What do you think is the right behavior?